File: Queue/LinkedQueue.c
#include<stdio.h>
#include <stdlib.h>

#define QUEUE_SIZE 4

typedef struct {
	int queue[QUEUE_SIZE];
	int front, rear;
} QueueType;

// 큐 생성
QueueType* createQueue() {
	QueueType* Q;
	Q = (QueueType*)malloc(sizeof(QueueType));
	Q->front = 0;	// front 초기값 설정
	Q->rear = 0;	// rear 초기값 설정
	return Q;
}
// 큐가 공백인지 확인
int isEmpty(QueueType* Q) {
	if (Q->front == Q->rear) {
		printf("Queue is Empty!");
		return 1;
	}
	return 0;
}

// 큐가 가득찼는지 확인
int isFull(QueueType *Q) {
	if (((Q->rear + 1) % QUEUE_SIZE) == Q->front) {
		printf("Queue is Full!");
		return 1;
	}
	return 0;
}

// 큐의 rear에 값을 삽입
void enQueue(QueueType* Q, int data) {
	if (isFull(Q)) return;
	Q->rear = (Q->rear + 1) % QUEUE_SIZE;
	Q->queue[Q->rear] = data;
}

// 큐의 front에 값을 삭제
int deQueue(QueueType* Q) {
	if (isEmpty(Q)) return 0;
	Q->front = (Q->front + 1) % QUEUE_SIZE;
	return Q->queue[Q->front];
}

int main(void) {
	QueueType* Q = createQueue();
	enQueue(Q, 10);
	enQueue(Q, 20);
	enQueue(Q, 30);
	printf("%d\n", deQueue(Q));
	printf("%d\n", deQueue(Q));
	printf("%d\n", deQueue(Q));
	return 0;
}

File: Queue/LinkedQueue.java
package Queue;

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

// 큐 클래스 정의
class LinkedListQueue<T> {
    private Node front;
    private Node rear;
    private int size;

    public LinkedListQueue() {
        this.front = null;
        this.rear = null;
        this.size = 0;
    }

    // 큐가 비어있는지 확인
    public boolean isEmpty() {
        return size == 0;
    }

    // 큐에 요소 추가
    public void enqueue(int data) {
        Node newNode = new Node(data);
        if (isEmpty()) {
            front = newNode;
        } else {
            rear.next = newNode;
        }
        rear = newNode;
        size++;
    }

    // 큐에서 요소 제거 및 반환
    public int dequeue() {
        if (isEmpty()) {
            throw new IllegalStateException("Queue is empty");
        }
        int data = front.data;
        front = front.next;
        if (front == null) {
            rear = null;
        }
        size--;
        return data;
    }
}

// 큐 사용 예제
public class LinkedQueue {
    public static void main(String[] args) {
        LinkedListQueue<Integer> queue = new LinkedListQueue<>();

        queue.enqueue(10);
        queue.enqueue(20);
        queue.enqueue(30);

        System.out.println(queue.dequeue());
        System.out.println(queue.dequeue());
        System.out.println(queue.dequeue());
    }
}